### ISSUE TITLE
"Not a directory (os error 20)" when trying to use boxxy to launch any program <autogenerated-issue>

### ISSUE THREAD

**Issue by hazelnot** [queer/boxxy#201]
## Error
```
Error: 
   0: Not a directory (os error 20)
```

## Metadata
|key|value|
|--|--|
|**version**|0.8.4|

**Comment by queer** [comment]
Please provide
- specific command run
- relevant parts of boxxy config

**Comment by hazelnot** [comment]
Specific command is `boxy war1gus`, though I also tried `boxxy dosbox-x`, since it autogenerated a DOSBox config and I thought I'd try one of the auto-generated ones as well to see if I got the same error.

Relevant parts are:
```
- name: DOSBox
  target: ~/.dosbox
  rewrite: ~/.config/dosbox
  mode: file
  context: []
  only: []
  env: {}
```
and
```
- name: stratagus
  target: ~/.stratagus
  rewrite: ~/.local/share/stratagus
```
though as I mentioned, the same thing seems to happen no matter what I try to run.

**Comment by hazelnot** [comment]
The culprit seems to have been one of the auto-generated rules. I deleted all but my custom rule and it suddenly works now. Not sure how to figure out which rule was the issue except deleting all 80-something of them one by one 😅.

**Comment by queer** [comment]
Can you try it with `boxxy -l debug ...`?

**Comment by hazelnot** [comment]
Ah I apologize, I saw this at a bad time and then forgot about it. I'll set a reminder for later cause I just saw this at a bad time again

**Comment by hazelnot** [comment]
```
$ boxxy -l debug ...
 DEBUG boxxy::config > loading rules (not asked not to!)
 DEBUG boxxy::config > searching for boxxy config starting at /home/hazelnot/git/Vanilla-Conquer/build
 DEBUG boxxy::config > checking for: /home/hazelnot/git/Vanilla-Conquer/build/boxxy.yaml
 DEBUG boxxy::config > checking for: /home/hazelnot/git/Vanilla-Conquer/boxxy.yaml
 DEBUG boxxy::config > checking for: /home/hazelnot/git/boxxy.yaml
 DEBUG boxxy::config > checking for: /home/hazelnot/boxxy.yaml
 DEBUG boxxy::config > checking for: /home/boxxy.yaml
 DEBUG boxxy::config > checking for: /boxxy.yaml
 DEBUG boxxy::config > ran out of parents to search!
 INFO  boxxy::config > loading rules from /home/hazelnot/.config/boxxy/boxxy.yaml
 INFO  boxxy::config > loaded 116 total rule(s)
 ERROR boxxy::config > command not found in $PATH or by path: ...
 DEBUG boxxy::config > searched $PATH: /home/hazelnot/.local/bin:/home/hazelnot/.nix-profile/bin:/nix/var/nix/profiles/default/bin:/usr/local/sbin:/usr/local/bin:/usr/bin:/var/lib/flatpak/exports/bin:/usr/lib/jvm/default/bin:/usr/bin/site_perl:/usr/bin/vendor_perl:/usr/bin/core_perl:/opt/rocm/bin:/usr/lib/rustup/bin
```

**Comment by queer** [comment]
Thank you 😊 I should be able to investigate this week.

**Comment by queer** [comment]
Sorry about the delay here 3: I have been dealing with some personal issues and will get to this when I have time.

**Comment by hazelnot** [comment]
No worries the one rule I added manually so far works fine, I can create some more probably without this happening as well 😅

**Comment by queer** [comment]
Finally able to work through the backlog of things :pensive:

If this is still relevant, does this only apply to running programs that have rules in the config file?

**Comment by jmickelin** [comment]
For what it's worth, I came upon this error message when I had initially not specified the `mode` correctly in my rule. I meant it to be in `file` mode, but since that stanza was missing and neither the target or the rewrite source existed yet on the system, the `boxxy`-invoked command happily created the target _as a directory_ (and then failed when trying to write in it). Something which I did not catch until much hair pulling later. On subsequent invocations, it was instead `boxxy` that failed because of the mismatch (I think?).

I'm not sure @hazelnot's case is at all similar, but I stumbled in here in frustration and wanted to leave a comment for posterity if nothing else.

An explicit check and more specific error message for this failure condition, where the rule says that the target is a file but it already exists as a directory, could be nice! 

Actually, what is the expected semantics when the target already exists on the physical system (either as the correct type or the wrong one)? Should it fail with an error or does it just shadow the file in the container? I expected the latter, since that's how bind mounts tend to work (I think?).

EDIT: After some further testing, I have to revise the above slightly. I think it _was_ actually `boxxy` that created the target directory, and the source _did_ exist (as a file) when I first ran it. Though, I'm a bit confused as to why. I thought the target should never exist outside of the container, but it does appear that target directories exist during the invocation of the program.

**Comment by queer** [comment]
@jmickelin Thank you for the detailed report!

> An explicit check and more specific error message for this failure condition, where the rule says that the target is a file but it already exists as a directory, could be nice!

I'll fix this and cut a new release soon!

> Should it fail with an error or does it just shadow the file in the container? I expected the latter, since that's how bind mounts tend to work (I think?).

It's supposed to shadow it.

> I thought the target should never exist outside of the container, but it does appear that target directories exist during the invocation of the program.

If the target file/directory does not exist, there needs to be a temporary file/directory created to bind-mount over. These are supposed to be automatically cleaned up after the program's execution ends, although some _older_ versions failed to do so correctly.

**Comment by queer** [comment]
Thank you both for the reports! This should be fixed in 0.8.5: https://github.com/queer/boxxy/releases/tag/v0.8.5

Please reopen this issue / open a new issue if you run into this or other issues!

**Comment by jmickelin** [comment]
Wow, that was fast :) Thank you so much!

> If the target file/directory does not exist, there needs to be a temporary file/directory created to bind-mount over. These are supposed to be automatically cleaned up after the program's execution ends, although some older versions failed to do so correctly.

Makes sense, good to know it's working as intended :+1:
